File: ts/instructions/createCollectionEntry.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface CreateCollectionEntryArgs {
  key: string
}

export interface CreateCollectionEntryAccounts {
  signer: PublicKey
  /** The entangler's state */
  state: PublicKey
  /** Mint fee */
  feeMint: PublicKey
  earner: PublicKey
  signerAccount: PublicKey
  earnerAccount: PublicKey
  /** The account storing the collection's data */
  entangledCollection: PublicKey
  /** The entangled collection entry */
  entangledCollectionEntry: PublicKey
  /** Common Solana programs */
  tokenProgram: PublicKey
  associatedTokenProgram: PublicKey
  systemProgram: PublicKey
  rent: PublicKey
}

export const layout = borsh.struct([borsh.str("key")])

/** Creates an entry in the collection map */
export function createCollectionEntry(
  args: CreateCollectionEntryArgs,
  accounts: CreateCollectionEntryAccounts
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.signer, isSigner: true, isWritable: true },
    { pubkey: accounts.state, isSigner: false, isWritable: false },
    { pubkey: accounts.feeMint, isSigner: false, isWritable: false },
    { pubkey: accounts.earner, isSigner: false, isWritable: true },
    { pubkey: accounts.signerAccount, isSigner: false, isWritable: true },
    { pubkey: accounts.earnerAccount, isSigner: false, isWritable: true },
    {
      pubkey: accounts.entangledCollection,
      isSigner: false,
      isWritable: false,
    },
    {
      pubkey: accounts.entangledCollectionEntry,
      isSigner: false,
      isWritable: true,
    },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    {
      pubkey: accounts.associatedTokenProgram,
      isSigner: false,
      isWritable: false,
    },
    { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.rent, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([167, 139, 193, 61, 122, 154, 187, 22])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      key: args.key,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId: PROGRAM_ID, data })
  return ix
}

File: programs/entangler/src/instructions/create_collection_entry.rs
use anchor_lang::prelude::*;
use anchor_spl::associated_token::AssociatedToken;
use anchor_spl::token::{transfer, Mint, Token, TokenAccount, Transfer};

use crate::seeds::{COLLECTION_ENTRY_SEED, COLLECTION_SEED, STATE_SEED};
use crate::state::{CollectionEntry, EntangledCollection, EntanglerState, MAX_KEY_SIZE};

pub fn create_collection_entry(ctx: Context<CreateCollectionEntry>, key: String) -> Result<()> {
    msg!("Creating the collection entry");

    let state = &mut ctx.accounts.state;
    let entangled_collection_entry = &mut ctx.accounts.entangled_collection_entry;
    entangled_collection_entry.id = ctx.accounts.entangled_collection.id.key();
    entangled_collection_entry.key = key;

    // Pay a fee to the DAO
    transfer(
        CpiContext::new(
            ctx.accounts.token_program.to_account_info(),
            Transfer {
                from: ctx.accounts.signer_account.to_account_info(),
                to: ctx.accounts.earner_account.to_account_info(),
                authority: ctx.accounts.signer.to_account_info(),
            },
        ),
        state.price,
    )?;

    Ok(())
}

#[derive(Accounts)]
#[instruction(key: String)]
pub struct CreateCollectionEntry<'info> {
    #[account(mut)]
    pub signer: Signer<'info>,

    /// The entangler's state
    #[account(
        seeds = [
            STATE_SEED.as_bytes(),
        ],
        bump,
        has_one = fee_mint,
    )]
    pub state: Box<Account<'info, EntanglerState>>,

    /// Mint fee
    #[account(address = state.fee_mint)]
    pub fee_mint: Account<'info, Mint>,

    /// CHECK: address
    #[account(
      mut,
    //   address = state.earner,
    )]
    pub earner: AccountInfo<'info>,

    #[account(
        init_if_needed,
        payer = signer,
        associated_token::mint = fee_mint,
        associated_token::authority = signer,
    )]
    pub signer_account: Box<Account<'info, TokenAccount>>,

    #[account(
        init_if_needed,
        payer = signer,
        associated_token::mint = fee_mint,
        associated_token::authority = earner,
    )]
    pub earner_account: Box<Account<'info, TokenAccount>>,

    /// The account storing the collection's data
    #[account(
        seeds = [
            COLLECTION_SEED.as_bytes(),
            &entangled_collection.id.to_bytes(),
        ],
        bump,
    )]
    pub entangled_collection: Account<'info, EntangledCollection>,

    /// The entangled collection entry
    #[account(
        init,
        space = CollectionEntry::LEN,
        payer = signer,
        seeds = [
            COLLECTION_ENTRY_SEED.as_bytes(),
            &key.as_bytes(),
        ],
        bump,
        constraint = key.len() <= MAX_KEY_SIZE,
        constraint = key.chars().all(|c| c.is_alphanumeric() || c == '_'),
    )]
    pub entangled_collection_entry: Box<Account<'info, CollectionEntry>>,

    /// Common Solana programs
    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
    pub rent: Sysvar<'info, Rent>,
}
